proxy/src/connection.rs

```diff
@@ -254,6 +249,39 @@ impl Future for ConditionallyUpgradeServerToTls {
     }
 }
 
+impl ConditionallyUpgradeServerToTlsInner {
+    /// Polls the underliyng socket for more data and buffers it.
+    ///
+    /// The buffer is matched for a TLS client hello message.
+    ///
+    /// An error is returned if the underlying socket has closed.
+    fn poll_match_client_hello(&mut self) -> Poll<tls::conditional_accept::Match, io::Error> {
+        let sz = try_ready!(self.socket.read_buf(&mut self.peek_buf));
+        if sz == 0 {
+            return Err(io::ErrorKind::BrokenPipe.into());
```

Comment:
Is it actually an error? It seems like various things like to open a single blank connection before scheduling new connections to do work...